### PR TITLE
[Utils] Add new --update-tests flag to llvm-lit

### LINKED ISSUE #81320: Incorporate update_*_test_checks.py into llvm-lit
Capturing the idea here from a discussion on Discord:
* It is annoying to have to specify tool paths to UTC scripts, but this is something llvm-lit has first-class knowledge of, so no more awkward `--llc-binary`/`--clang`
* llvm-lit knows how to do test discovery, parallel job dispatch, and gives a nice progress bar when you use `-sv`

### PR DESCRIPTION
**Pull request by hnrklssn**
This adds a flag to lit for detecting and updating failing tests when possible to do so automatically. The flag uses a plugin architecture where config files can add additional auto-updaters for the types of tests in the test suite. When a test fails with `--update-tests` enabled lit passes the test RUN invocation and output to each registered test updater until one of them signals that it updated the test (or all test updaters have been run). As such it is the responsibility of the test updater to only update tests where it is reasonably certain that it will actually fix the test, or come close to doing so.

Initially adds support for UpdateVerifyTests and UpdateTestChecks. The flag is currently only implemented for lit's internal shell, so `--update-tests` implies `LIT_USE_INTERNAL_SHELL=1`.

Builds on work in #97369
Fixes #81320

### CODE REVIEW

**Comment by hnrklssn** [comment]
Note that the first 4 commits are from https://github.com/llvm/llvm-project/pull/97369; only the last 2 are new. I'll merge that tomorrow when I have more time to be ready to revert in case something should happen.

File: clang/test/Sema/lit.local.cfg
Comment (nhaehnle) [inline clang/test/Sema/lit.local.cfg:13]: I believe this can be simplified to `os.path.join(clangdir, "..", "..", "utils")`?

File: clang/test/utils/update-verify-tests/Inputs/duplicate-diag.c
Comment (nhaehnle) [inline clang/test/utils/update-verify-tests/Inputs/duplicate-diag.c:2]: I still think those new invocations of `diff` should be in `lit-plugin.test`.

Those `diff` lines attempt to check that this file (or rather its temporarily copied version) was overwritten correctly.

It isn't sound to have the code to check whether a file was correctly overwritten in the file that is overwritten.

**Review by nhaehnle — COMMENTED** [review]
Thanks for splitting this up. It does look mostly good to me, but I do have two comments.

File: clang/test/Sema/lit.local.cfg
Comment (hnrklssn) [reply clang/test/Sema/lit.local.cfg]: Ah neat, wasn't aware that was a thing

**Review by nhaehnle — COMMENTED** [review]
Thank you. This pretty much LGTM, but it just occurred to me that there's a docs/CommandGuide/lit.rst which should be updated to document the new option.

**Comment by erichkeane** [comment]
I mentioned on a separate commit here, but as a reviewer/code-owner, I very much dislike the existence of tooling like this. I believe it to be actively harmful to the project.

**Comment by nikic** [comment]
As update-verify-tests was reverted, can you rebase this to just the update_test_checks support? On the LLVM side, we definitely want to have this feature.

**Comment by hnrklssn** [comment]
> Thank you. This pretty much LGTM, but it just occurred to me that there's a docs/CommandGuide/lit.rst which should be updated to document the new option.

Done

> As update-verify-tests was reverted, can you rebase this to just the update_test_checks support? On the LLVM side, we definitely want to have this feature.

and done

**Review by nikic — APPROVED** [review]
LGTM. I gave this a try and it seems to work well!

**Pull request merged**
